File: merrrt_visualization/src/merrrt_visualization/rviz_animation_controller.cpp
#include <merrrt_visualization/rviz_animation_controller.h>
#include <peter_msgs/GetAnimControllerState.h>
#include <peter_msgs/GetFloat32.h>
#include <peter_msgs/AnimationControl.h>
#include <std_msgs/Empty.h>

#include <QApplication>
#include <QCheckBox>
#include <QDoubleSpinBox>
#include <QMainWindow>
#include <QPushButton>
#include <iostream>

#define create_service_options(type, name, bind)                                                                       \
  ros::AdvertiseServiceOptions::create<type>(name, bind, ros::VoidPtr(), &queue_)

namespace merrrt_visualization
{
RVizAnimationController::RVizAnimationController(QWidget *parent) : rviz::Panel(parent)
{
  ui.setupUi(this);
  connect(ui.forward_button, &QPushButton::clicked, this, &RVizAnimationController::ForwardClicked);
  connect(ui.backward_button, &QPushButton::clicked, this, &RVizAnimationController::BackwardClicked);
  connect(ui.play_forward_button, &QPushButton::clicked, this, &RVizAnimationController::PlayForwardClicked);
  connect(ui.play_backward_button, &QPushButton::clicked, this, &RVizAnimationController::PlayBackwardClicked);
  connect(ui.pause_button, &QPushButton::clicked, this, &RVizAnimationController::PauseClicked);
  connect(ui.done_button, &QPushButton::clicked, this, &RVizAnimationController::DoneClicked);
  connect(ui.loop_checkbox, &QCheckBox::toggled, this, &RVizAnimationController::LoopToggled);
  connect(ui.auto_play_checkbox, &QCheckBox::toggled, this, &RVizAnimationController::AutoPlayToggled);
  connect(ui.period_spinbox, qOverload<double>(&QDoubleSpinBox::valueChanged), this, &RVizAnimationController::PeriodChanged);

  command_pub_ = ros_node_.advertise<peter_msgs::AnimationControl>("rviz_anim/control", 10);


  auto get_state_bind = [this](peter_msgs::GetAnimControllerStateRequest &req,
                               peter_msgs::GetAnimControllerStateResponse &res)
  {
    (void) req; // unused
    res.state.auto_play = ui.auto_play_checkbox->isChecked();
    res.state.loop = ui.loop_checkbox->isChecked();
    res.state.period = static_cast<float>(ui.period_spinbox->value());
    return true;
  };
  auto get_state_so = create_service_options(peter_msgs::GetAnimControllerState, "rviz_anim/get_state", get_state_bind);
  get_state_srv_ = ros_node_.advertiseService(get_state_so);

  // this is stupid why must I list this type here but not when I do this for services!?
  boost::function<void(const std_msgs::Int64::ConstPtr &)> time_cb = [this](const std_msgs::Int64::ConstPtr &msg)
  {
    TimeCallback(msg);
  };
  auto time_sub_so = ros::SubscribeOptions::create("rviz_anim/time", 10, time_cb, ros::VoidPtr(), &queue_);
  time_sub_ = ros_node_.subscribe(time_sub_so);

  // this is stupid why must I list this type here but not when I do this for services!?
  boost::function<void(const std_msgs::Int64::ConstPtr &)> max_time_cb = [this](const std_msgs::Int64::ConstPtr &msg)
  {
    MaxTimeCallback(msg);
  };
  auto max_time_sub_so = ros::SubscribeOptions::create("rviz_anim/max_time", 10, max_time_cb, ros::VoidPtr(), &queue_);
  max_time_sub_ = ros_node_.subscribe(max_time_sub_so);

  ros_queue_thread_ = std::thread([this] { QueueThread(); });
}

RVizAnimationController::~RVizAnimationController()
{
  queue_.clear();
  queue_.disable();
  ros_node_.shutdown();
  ros_queue_thread_.join();
}

void RVizAnimationController::TimeCallback(const std_msgs::Int64::ConstPtr &msg)
{
  QString text;
  text.sprintf("%3ld", msg->data);
  ui.step_number_label->setText(text);
  update();
}

void RVizAnimationController::MaxTimeCallback(const std_msgs::Int64::ConstPtr &msg)
{
  auto const text = QString::number(msg->data);
  ui.max_step_number_label->setText(text);
  update();
}

void RVizAnimationController::DoneClicked()
{
  peter_msgs::AnimationControl cmd;
  cmd.command = peter_msgs::AnimationControl::DONE;
  command_pub_.publish(cmd);
}

void RVizAnimationController::ForwardClicked()
{
  peter_msgs::AnimationControl cmd;
  cmd.command = peter_msgs::AnimationControl::STEP_FORWARD;
  command_pub_.publish(cmd);
}

void RVizAnimationController::BackwardClicked()
{
  peter_msgs::AnimationControl cmd;
  cmd.command = peter_msgs::AnimationControl::STEP_BACKWARD;
  command_pub_.publish(cmd);
}

void RVizAnimationController::PauseClicked()
{
  peter_msgs::AnimationControl cmd;
  cmd.command = peter_msgs::AnimationControl::PAUSE;
  command_pub_.publish(cmd);
}

void RVizAnimationController::PlayForwardClicked()
{
  peter_msgs::AnimationControl cmd;
  cmd.command = peter_msgs::AnimationControl::PLAY_FORWARD;
  command_pub_.publish(cmd);
}

void RVizAnimationController::PlayBackwardClicked()
{
  peter_msgs::AnimationControl cmd;
  cmd.command = peter_msgs::AnimationControl::PLAY_BACKWARD;
  command_pub_.publish(cmd);
}

void RVizAnimationController::LoopToggled()
{
  peter_msgs::AnimationControl cmd;
  cmd.state.loop = ui.loop_checkbox->isChecked();
  cmd.command = peter_msgs::AnimationControl::SET_LOOP;
  command_pub_.publish(cmd);

  if (ui.auto_play_checkbox->isChecked()) {
    ROS_WARN_STREAM("Auto-play takes precedence over looping");
    // show a warning here
  }
}

void RVizAnimationController::AutoPlayToggled()
{
  peter_msgs::AnimationControl cmd;
  cmd.state.auto_play = ui.auto_play_checkbox->isChecked();
  cmd.command = peter_msgs::AnimationControl::SET_AUTO_PLAY;
  command_pub_.publish(cmd);

  if (ui.loop_checkbox->isChecked()) {
    ROS_WARN_STREAM("Looping takes precedence over auto-play");
    // show a warning here
  }
}


void RVizAnimationController::PeriodChanged(double period)
{
  peter_msgs::AnimationControl cmd;
  cmd.state.period = period;
  cmd.command = peter_msgs::AnimationControl::SET_PERIOD;
  command_pub_.publish(cmd);
}

void RVizAnimationController::QueueThread()
{
  double constexpr timeout = 0.01;
  while (ros_node_.ok())
  {
    queue_.callAvailable(ros::WallDuration(timeout));
  }
}

}  // namespace merrrt_visualization

#include <pluginlib/class_list_macros.h>

PLUGINLIB_EXPORT_CLASS(merrrt_visualization::RVizAnimationController, rviz::Panel)

File: merrrt_visualization/src/merrrt_visualization/rviz_animation_controller.py
from time import sleep
from typing import Dict, List, Callable

import numpy as np

import rospy
from link_bot_pycommon.experiment_scenario import ExperimentScenario
from moonshine.moonshine_utils import numpify
from peter_msgs.msg import AnimationControl
from peter_msgs.srv import GetAnimControllerStateRequest, GetAnimControllerState
from std_msgs.msg import Int64


class RvizAnimationController:

    def __init__(self, time_steps=None, n_time_steps: int = None):
        if time_steps is None and n_time_steps is None:
            raise ValueError("you have to pass either n_time_steps or time_steps")
        if time_steps is not None:
            self.time_steps = np.array(time_steps, dtype=np.int64)
        if n_time_steps is not None:
            self.time_steps = np.arange(n_time_steps, dtype=np.int64)
        self.command_sub = rospy.Subscriber("/rviz_anim/control", AnimationControl, self.on_control)
        self.time_pub = rospy.Publisher("/rviz_anim/time", Int64, queue_size=10)
        self.max_time_pub = rospy.Publisher("/rviz_anim/max_time", Int64, queue_size=10)
        get_srv_name = "/rviz_anim/get_state"
        self.get_state_srv = rospy.ServiceProxy(get_srv_name, GetAnimControllerState)

        rospy.logdebug(f"waiting for {get_srv_name}")
        rospy.wait_for_service(get_srv_name)
        rospy.logdebug(f"connected.")

        self.idx = 0
        self.max_idx = self.time_steps.shape[0]
        self.max_t = self.time_steps[-1]
        state_res = self.get_state_srv(GetAnimControllerStateRequest())
        self.auto_play = state_res.state.auto_play
        self.loop = state_res.state.loop
        self.period = state_res.state.period
        self.playing = self.auto_play or self.loop
        self.should_step = False
        self.fwd = True
        self.done = False

    def on_control(self, msg: AnimationControl):
        if msg.command == AnimationControl.STEP_BACKWARD:
            self.on_bwd()
        elif msg.command == AnimationControl.STEP_FORWARD:
            self.on_fwd()
        elif msg.command == AnimationControl.PLAY_BACKWARD:
            self.on_play_backward()
        elif msg.command == AnimationControl.PLAY_FORWARD:
            self.on_play_forward()
        elif msg.command == AnimationControl.PAUSE:
            self.on_pause()
        elif msg.command == AnimationControl.DONE:
            self.on_done()
        elif msg.command == AnimationControl.SET_LOOP:
            self.loop = msg.state.loop
        elif msg.command == AnimationControl.SET_AUTO_PLAY:
            self.loop = msg.state.loop
        elif msg.command == AnimationControl.SET_PERIOD:
            self.period = msg.state.period
        else:
            raise NotImplementedError(f"Unsupported animation control {msg.command}")

    def on_fwd(self):
        self.should_step = True
        self.playing = False
        self.fwd = True

    def on_bwd(self):
        self.should_step = True
        self.playing = False
        self.fwd = False

    def on_play_forward(self):
        self.playing = True
        self.fwd = True

    def on_play_backward(self):
        self.playing = True
        self.fwd = False

    def on_pause(self):
        self.playing = False

    def on_done(self):
        self.done = True

    def step(self):
        if self.playing:
            # don't use ros time because we don't want to rely on simulation time
            sleep(self.period)
        else:
            while not self.should_step and not self.playing and not self.done:
                sleep(0.01)

        if self.fwd:
            if self.idx < self.max_idx - 1:
                self.idx += 1
            else:
                if self.auto_play:
                    self.done = True
                elif self.loop:
                    self.idx = 0
                else:
                    self.playing = False
        else:
            if self.idx > 0:
                self.idx -= 1

        t_msg = Int64()
        t_msg.data = self.time_steps[self.idx]
        self.time_pub.publish(t_msg)

        self.should_step = False

        max_t_msg = Int64()
        max_t_msg.data = self.time_steps[-1]
        self.max_time_pub.publish(max_t_msg)

        return self.done

    def t(self):
        return self.time_steps[self.idx]


class RvizSimpleStepper:

    def __init__(self):
        self.command_sub = rospy.Subscriber("/rviz_anim/control", AnimationControl, self.on_control)
        self.should_step = False
        self.play = False

    def on_control(self, msg: AnimationControl):
        if msg.command == AnimationControl.STEP_FORWARD:
            self.should_step = True
        elif msg.command == AnimationControl.PLAY_FORWARD:
            self.should_step = True
            self.play = True
        elif msg.command == AnimationControl.PAUSE:
            self.play = False

    def step(self):
        while not self.should_step:
            sleep(0.05)
        if not self.play:
            self.should_step = False


# pylint: disable=too-few-public-methods
class RvizAnimation:

    def __init__(self,
                 scenario: ExperimentScenario,
                 n_time_steps: int,
                 init_funcs: List[Callable],
                 t_funcs: List[Callable]):
        self.scenario = scenario
        self.init_funcs = init_funcs
        self.t_funcs = t_funcs
        self.n_time_steps = n_time_steps

    def play(self, example: Dict):
        example = numpify(example)
        for init_func in self.init_funcs:
            init_func(self.scenario, example)

        controller = RvizAnimationController(n_time_steps=self.n_time_steps)
        while not controller.done:
            t = controller.t()

            for t_func in self.t_funcs:
                t_func(self.scenario, example, t)

            controller.step()
